operator-common/src/main/java/io/strimzi/operator/common/operator/resource/ResourceSupport.java

```diff
@@ -0,0 +1,165 @@
+     */
+    public <T, U> Future<U> selfClosingWatch(Watchable<Watch, Watcher<T>> watchable,
+                                             BiFunction<Watcher.Action, T, U> watchFn,
+                                             long timeoutMs) {
+
+        return new Watcher<T>() {
+            private final Future<Watch> watchFuture;
+            private final Future<U> doneFuture;
+            private final Future<U> resultFuture;
+            private final long timerId;
+
+            /* init */
+            {
+                this.watchFuture = Future.future();
+                this.doneFuture = Future.future();
+                this.resultFuture = Future.future();
+                this.timerId = vertx.setTimer(timeoutMs, ignored -> {
+                    doneFuture.tryFail(new TimeoutException());
+                });
+                CompositeFuture.join(watchFuture, doneFuture).setHandler(joinResult -> {
+                    Future<Void> closeFuture;
+                    if (watchFuture.succeeded()) {
+                        closeFuture = closeOnWorkerThread(watchFuture.result());
+                    } else {
+                        closeFuture = Future.succeededFuture();
+                    }
+                    closeFuture.setHandler(closeResult -> {
+                        vertx.runOnContext(ignored2 -> {
+                            LOGGER.warn("Completing watch future");
+                            if (joinResult.succeeded() && closeResult.succeeded()) {
+                                resultFuture.complete(joinResult.result().resultAt(1));
+                            } else {
+                                resultFuture.fail(collectCauses(joinResult, closeResult));
+                            }
+                        });
+                    });
+                });
+                Watch watch = watchable.watch(this);
+                LOGGER.debug("Opened watch {}", watch);
+                watchFuture.complete(watch);
+            }
+
+            @Override
+            public void eventReceived(Action action, T resource) {
+                vertx.<U>executeBlocking(
+                    f -> {
+                        try {
+                            U apply = watchFn.apply(action, resource);
+                            if (apply != null) {
+                                f.tryComplete(apply);
+                                vertx.cancelTimer(timerId);
+                            } else {
+                                LOGGER.debug("Not yet complete");
+                            }
+                        } catch (Throwable t) {
+                            f.tryFail(t);
+                        }
+                    },
+                    true,
+                    ar -> {
+                        doneFuture.handle(ar);
+                    });
+            }
+
+            @Override
+            public void onClose(KubernetesClientException cause) {
```

Comment:
Should we at least log the exception? Or are we sure that it can be always ignored?